presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -546,19 +548,38 @@ protected Object visitCoalesceExpression(CoalesceExpression node, Object context
             List<Object> values = node.getOperands().stream()
                     .map(value -> processWithExceptionHandling(value, context))
                     .filter(Objects::nonNull)
+                    .flatMap(expression -> {
+                        if (expression instanceof CoalesceExpression) {
+                            return ((CoalesceExpression) expression).getOperands().stream();
+                        }
+                        return Stream.of(expression);
+                    })
                     .collect(Collectors.toList());
 
             if ((!values.isEmpty() && !(values.get(0) instanceof Expression)) || values.size() == 1) {
                 return values.get(0);
             }
-
-            List<Expression> expressions = values.stream()
-                    .map(value -> toExpression(value, type))
-                    .collect(Collectors.toList());
+            ImmutableList.Builder<Expression> operandsBuilder = ImmutableList.builder();
+            Set<Expression> visitedExpression = new HashSet<>();
+            for (Object value : values) {
+                Expression expression = toExpression(value, type);
+                if (!visitedExpression.contains(expression)) {
+                    operandsBuilder.add(expression);
+                    visitedExpression.add(expression);
```

Comment:
I think we should not dedup non-deterministic functions, see: https://github.com/prestodb/presto/issues/7374#issuecomment-280138304

Reply:
But is `coalsce` is a deterministic function right ? Would we face any issue if we are dedup a deterministic function ?

Reply:
`coalesce` is deterministic, but operands not necessarily.

Reply:
`coalesce` is a language construct, not a function, since its arguments are evaluated lazily (the SQL standard defines it in terms of a `CASE` expression).

Reply:
> since its arguments are evaluated lazily

Actually, the spec doesn't say anything about lazy evaluation -- only about the order of evaluation. I.e., it returns the `<result>` of the *first* `<search when clause>` for which the `<search condition>` evaluates to true. So:

```
case 
   when true then 1
   when true then 2
end
```

is specified to return 1, and

```
case 
   when true then 2
   when true then 1
end
```

is specified to return 2.